python/rockets/client.py

```diff
@@ -0,0 +1,136 @@
+        """
+        if not loop:
+            loop = asyncio.get_event_loop()
+
+        if loop.is_running():
+            thread_loop = asyncio.new_event_loop()
+
+            def _start_background_loop(loop):
+                asyncio.set_event_loop(loop)
+                loop.run_forever()
+
+            self._thread = Thread(target=_start_background_loop, args=(thread_loop,))
+            self._thread.daemon = True
+            self._thread.start()
+
+            self._client = AsyncClient(url, subprotocols=subprotocols, loop=thread_loop)
+            self._async_client = AsyncClient(url, subprotocols=subprotocols, loop=loop)
+        else:
+            self._thread = None
+            self._client = AsyncClient(url, subprotocols=subprotocols, loop=loop)
+            self._async_client = self._client
+
+    @copydoc(AsyncClient.url)
+    def url(self):  # noqa: D102 pylint: disable=missing-docstring
+        return self._client.url()
+
+    @copydoc(AsyncClient.connected)
+    def connected(self):  # noqa: D102 pylint: disable=missing-docstring
+        return self._client.connected()
+
+    @copydoc(AsyncClient.connect)
+    def connect(self):  # noqa: D102 pylint: disable=missing-docstring
+        self._verify_environment()
+        self._call_sync(self._client.connect())
+
+    @copydoc(AsyncClient.disconnect)
+    def disconnect(self):  # noqa: D102 pylint: disable=missing-docstring
+        self._verify_environment()
+        self._call_sync(self._client.disconnect())
+
+    @copydoc(AsyncClient.as_observable)
+    def as_observable(self):  # noqa: D102 pylint: disable=missing-docstring
+        return self._client.as_observable()
+
+    @copydoc(AsyncClient.send)
+    def send(self, message):  # noqa: D102 pylint: disable=missing-docstring
+        self._verify_environment()
+        self._call_sync(self._client.send(message))
+
+    @copydoc(AsyncClient.notify)
+    def notify(self, method, params=None):  # noqa: D102 pylint: disable=missing-docstring
+        self._verify_environment()
+        self._call_sync(self._client.notify(method, params))
+
+    @copydoc(AsyncClient.request)
+    def request(self, method, params=None, response_timeout=None):  # noqa: D102,D205 pylint: disable=C0111,W9011,W9012,W9015,W9016
+        """
+        :param int response_timeout: number of seconds to wait for the response
+        :raises TimeoutError: if request was not answered within given response_timeout
+        """
+        self._verify_environment()
+        return self._call_sync(self._client.request(method, params), response_timeout)
+
+    @copydoc(AsyncClient.batch)
+    def batch(self, requests, response_timeout=None):  # noqa: D102,D205 pylint: disable=C0111,W9011,W9012,W9015,W9016
+        """
+        :param int response_timeout: number of seconds to wait for the response
+        :raises TimeoutError: if request was not answered within given response_timeout
+        """
+        self._verify_environment()
```

Comment:
I think self._verify_environment() could be moved inside _call_sync() since it's called everytime

Reply:
There was a reason I had to separate that at one point during development, but apparently that's no longer true.